src/main/java/io/jenkins/plugins/analysis/core/steps/PublishIssuesStep.java

```diff
@@ -572,7 +576,13 @@ protected ResultAction run() throws IOException, InterruptedException, IllegalSt
             IssuesPublisher publisher = new IssuesPublisher(getRun(), report, healthDescriptor, qualityGate,
                     name, referenceJobName, ignoreQualityGate, ignoreFailedBuilds,
                     getCharset(sourceCodeEncoding), getLogger(report));
-            return publisher.attachAction();
+            ResultAction resultAction = publisher.attachAction();
+            QualityGateStatus status = resultAction.getResult().getQualityGateStatus();
+            if (status == QualityGateStatus.WARNING) {
+                WarningAction action = new WarningAction("Some quality gates have been missed: overall result is " + status);
+                getContext().get(FlowNode.class).addOrReplaceAction(action);
+            }
+            return resultAction;
```

Comment:
I think it makes more sense if this code is called in `IssuePublisher`, since most users (especially for declarative pipeline) use the `IssueRecorder` step. 
So here: https://github.com/jenkinsci/warnings-ng-plugin/blob/master/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java#L145
Or here: https://github.com/jenkinsci/warnings-ng-plugin/blob/master/src/main/java/io/jenkins/plugins/analysis/core/util/QualityGateStatus.java#L58
I don't know if this FlowNode is available for a `SimpleBuildStep`?

Reply:
Yeah, that was what I had in mind with this comment from the PR description:

> Another idea I had was to create some kind of `ResultHandler` class that can be passed to `IssuePublisher` and then to a new `QualityGateStatus.setResult(ResultHandler)` method, but I wasn't sure if we wanted to make those kinds of changes if `publishIssues` is the only caller that would actually use `ResultHandler`.

> I don't know if this FlowNode is available for a SimpleBuildStep

Unfortunately no, I don't think there is any way to make this work without changing the API of `SimpleBuildStep` (and I don't think that is likely to happen). When used in a Pipeline, `SimpleBuildSteps` are actually [executed by `CoreStep`](https://github.com/jenkinsci/workflow-basic-steps-plugin/blob/5d48ed444cd4041d4b00a2c7197559406d1a2c9a/src/main/java/org/jenkinsci/plugins/workflow/steps/CoreStep.java#L80), so they have no way to get the current `FlowNode` because they don't have access to any Pipeline or Step-specific APIs. (We _could_ add a compile-scope dependency on workflow-job and check `run instanceof WorkflowRun`, then use `WorkflowRun.getExecution().getCurrentHeads()` to try to find a `FlowNode` for `CoreStep` where `CoreStep.delegate == this`, but I'd consider that a serious hack.)

Reply:
Hmm, that would prevent all plugins that just implement `SimpleBuildStep` (there are a lot of them) not aware of the new concept :cry:

Maybe it would make sense to provide an additional action `SimpleBuildStep2` that provides the missing parameter `FlowNode`?

Reply:
Well, I think the idea behind `SimpleBuildStep` is that it can work in various kinds of jobs, so it can't have any behaviors that are specific to a certain job type (and it is defined in core, so I'm not sure we even could provide a `SimpleBuildStep2` that takes a `FlowNode` as a parameter. Perhaps we could provide a new overload of `SimpleBuildStep.perform` that returns some composite type with a result and a message, and handle that in `CoreStep`, but I'm not sure we want to create a new core API just for this). 

I guess the crux of the issue is that tracking per-step failures is very much a Pipeline-specific behavior, but it seems like you are using `SimpleBuildStep` intentionally to avoid needing to create a separate step just for Pipeline since the majority of the behavior is agnostic to the specifics of the build. I'm not sure if we can do anything easily to fix that without either a core API or creating a new step here that uses `IssueRecorder` internally (perhaps a `Step` with the same symbol name as `IssueRecorder` would take precedence when evaluated in Pipeline allowing transparent migration for users?).

Reply:
I see. Creating another step seems like a simpler solution. I already extracted the behavior of `IssueRecorder` and `ScanForIssuesStep` and `PublishIssuesStep` into separate classes in order to avoid duplication. Adding another step would not harm much (a lot of duplication will be required for data-binding of the fields though). 

> (perhaps a Step with the same symbol name as IssueRecorder would take precedence when evaluated in Pipeline allowing transparent migration for users?)

Shouldn't moving the symbol from `IssueRecorder` to a new step should do the trick?

Reply:
> Shouldn't moving the symbol from IssueRecorder to a new step should do the trick?

Yeah that would be the easiest approach as long as the only uses of the symbol today are in Pipeline and so we wouldn't break anything for anyone else.

Reply:
Ok, I've just force-pushed with a new commit that introduces `RecordIssuesStep`, which is related to `IssuesRecorder` similarly to the way `PublishIssuesStep` relates to `IssuesPublisher`. Care to take a look to see if the overall approach seems ok to you? (I'd still need to finalize a few things like jelly files for the new step, but figured I'd wait to do that unless you agree that the approach seems ok).

Reply:
I think it would be better if we would rather have a new Interface that has two different implementations: one sets the status of the `Run` and the other additionally sets the status of the `FlowNode`. Then we can avoid `null` invocations.

Reply:
The overall approach of copying the recorder to a new step class looks ok. I think we can't avoid the duplication here (this is a similar and ugly thing that I already had with the maven plugin support and the scan and publish steps).